### PR TITLE
Update mypy config and make it work well with pre-commit

### PR DESCRIPTION
**Pull request by mathbunnyru**
What I've done:
- Updated mypy config to use `--strict --follow-imports error`. This makes this tool much more efficient in finding errors.
- Fixed all the errors
- Made `pre-commit` mypy hook use the same config
- Made sure the `pre-commit` hook works well and disabled `mypy` for `git commit` stage
- Added some inline comments to explain why and how I did it.

### CODE REVIEW

**Comment by mathbunnyru** [comment]
Unfortunately, `pre-commit` mypy doesn't work the same way as explicit mypy run.
So, it's better to run `mypy --config-file ./mypy.ini ./` for now.
CC: @trallard

File: mypy.ini [hunk 1]
```diff
@@ -4,6 +4,9 @@ follow_imports = normal
 strict = False
 no_incremental = True
 
+[mypy-Cython.*]
```
Comment (consideRatio) [inline mypy.ini:7]: Can you add a comment to the top of this config file saying what we use it for, and what mypy is?

**Comment by consideRatio** [comment]
> Unfortunately, `pre-commit` mypy doesn't work the same way as explicit mypy run. So, it's better to run `mypy --config-file ./mypy.ini ./` for now.

If we cant use if with pre-commit then that is also relevance to clarify in the file as a comment/fixme preferably linking out to something relevant to track if its to be fixed over time - a github issue in this or upstream repo.

**Comment by consideRatio** [comment]
> Unfortunately, `pre-commit` mypy doesn't work the same way as explicit mypy run.
> So, it's better to run `mypy --config-file ./mypy.ini ./` for now.

If it doesn't work the same way, and it's not of importance, I suggest we stick with anything that can be done with pre-commit instead of sticking with an option that requires manual intervention etc. I've not read up on mypy or what it does though.

File: mypy.ini (hunk 1 above)
Comment (mathbunnyru) [reply mypy.ini:21]: Thanks!
I added the docs to make it clear.

**Comment by mathbunnyru** [comment]
@consideRatio thank you for very good suggestions.
I think I made it work well for `pre-commit` and documented the workaround.
Please, take a look and tell me what you think.

**Review by consideRatio — APPROVED** [review]
Beautiful @mathbunnyru! Thanks for being so thorough with the transition to make this code typed, with all this setup I think this repo is really set to maintain typings!!! :heart: :tada:

**Pull request merged**